### DISCUSSION TITLE
About Equals Operator Overloading in Interface: What Constraint Does CS9046 Need?

### DISCUSSION THREAD

**Discussion by Orange23333** [dotnet/csharplang#7330] [Q&A] [answered]
# Problem
```csharp
public interface IPathNode<TValue, TCoordinate> where TValue : IBinaryNumber<TValue> where TCoordinate : struct
{
    Guid NId { get; }

    public virtual static bool operator ==<TSelf>(TSelf left, TSelf right) where TSelf : IPathNode<TValue, TCoordinate>
    {
        return left.NId == right.NId;
    }	
    public virtual static bool operator !=<TSelf>(TSelf left, TSelf right) where TSelf : IPathNode<TValue, TCoordinate>
    {
        return left.NId != right.NId;
    }
}
```
![Screenshot of Code Happened CS9046](https://github.com/dotnet/csharplang/assets/18095289/3130762b-a7d1-4360-b095-2908265dba1a)
According to the picture above. It said the type parameter needs constraint. But how?
* This is an operator overriding method in interface, but not a generic method.
* I think really can't constrain anything here with `where` any more.

# Blob
* I tried to overload `+`, `>` and `<`. No error comes out. Yes, the `CS9046` mentions only `==` and `!=`. No loss is [highlighted](https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/operator-overloads#overloading-operator-). So special (with an I don't know why error) that really confused me.
* No such document about `CS9046` on MSDN so far.
* I tried to search `C# operator overloading needs constraint in interface` on Bing. There are many pages about operator overloading, generics constraint, operator overloading of generics with `+`, an [issue](https://github.com/dotnet/roslyn/issues/5624) with similar title but for `+` as well. But there is nothing about `==`.

---

I need help!

**Comment by CyrusNajmabadi** [comment, marked as answer]
the pattern this is looking for is something like:

```c#
interface I<X> where X : I<X>
{
     public virtual static operator ==(X x, ...)
}
```

here, one of the parameters of the equality operator is a type parameter on I<X>, (specifically `X`) and that type parameter itself is constrained to `I<X>`.

**Reply by Orange23333** [reply]
Thanks!
Looks like we can't escape `TSelf` in this version. So, we have to enjoy the advantages that it brings.

---

And I just found this [Type arguments implement declared interface](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/generics/constraints-on-type-parameters#type-arguments-implement-declared-interface) after your answer was given out. A `TSelf` really gives me many conveniences.
According to my test and the column("Type arguments implement declared interface"), `public virtual static I operator +(I left, ...)`is allowed (although it needs an explicit interface implementation). But I still haven't understood why `==` is not allowed.
